prow/cmd/branchprotector/protect.go

```diff
@@ -0,0 +1,293 @@
+	"strings"
+	"sync"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+)
+
+type options struct {
+	config   string
+	token    string
+	confirm  bool
+	endpoint string
+}
+
+func jobRequirements(jobs []config.Presubmit, after bool) []string {
+	var required []string
+	for _, j := range jobs {
+		if !after && (!j.AlwaysRun || j.SkipReport) {
+			continue
+		}
+		required = append(required, j.Context)
+		required = append(required, jobRequirements(j.RunAfterSuccess, true)...)
+	}
+	return required
+}
+
+func repoRequirements(org, repo string, cfg config.Config) []string {
+	p, ok := cfg.Presubmits[org+"/"+repo]
+	if !ok {
+		return nil
+	}
+	return jobRequirements(p, false)
+}
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.endpoint, "endpoint", "https://api.github.com", "Github api endpoint, may differ for enterprise")
+	flag.StringVar(&o.config, "config", "", "Path to prow config.yaml")
+	flag.StringVar(&o.token, "token", "", "Path to github token")
+	flag.BoolVar(&o.confirm, "confirm", false, "Mutate github if set")
+	flag.Parse()
+	return o
+}
+
+type Requirements struct {
+	Org      string
+	Repo     string
+	Branch   string
+	Protect  bool
+	Contexts []string
+	Pushers  []string
+}
+
+type Errors struct {
+	lock sync.Mutex
+	errs []error
+}
+
+func (e *Errors) add(err error) {
+	e.lock.Lock()
+	log.Println(err)
+	defer e.lock.Unlock()
+	e.errs = append(e.errs, err)
+}
+
+func main() {
+	o := flagOptions()
+	if o.config == "" {
+		log.Fatal("empty --config")
+	}
+
+	cfg, err := config.Load(o.config)
+	if err != nil {
+		log.Fatalf("Failed to load --config=%s: %v", o.config, err)
+	}
+
+	if o.token == "" {
+		log.Fatal("empty --token")
+	}
+	b, err := ioutil.ReadFile(o.token)
+	if err != nil {
+		log.Fatalf("cannot read --token: %v", err)
+	}
+	_, err = url.Parse(o.endpoint)
+	if err != nil {
+		log.Fatalf("Must specify valid --endpoint URL: %v", err)
+	}
+
+	var c *github.Client
+	tok := strings.TrimSpace(string(b))
+	if o.confirm {
+		c = github.NewClient(tok, o.endpoint)
+	} else {
+		c = github.NewDryRunClient(tok, o.endpoint)
+	}
+	c.Throttle(300, 100) // 300 hourly tokens, bursts of 100
+
+	p := Protector{
+		client:         c,
+		cfg:            cfg,
+		updates:        make(chan Requirements),
+		errors:         Errors{},
+		completedRepos: make(map[string]bool),
+		done:           make(chan []error),
+	}
+
+	go p.ConfigureBranches()
+	p.Protect()
+	close(p.updates)
+	errors := <-p.done
+	if len(errors) > 0 {
+		log.Fatalf("Encountered %d errrors protecting branches: %v", len(errors), errors)
+	}
+}
+
+type client interface {
+	RemoveBranchProtection(org, repo, branch string) error
+	UpdateBranchProtection(org, repo, branch string, contexts, pushers []string) error
+	GetBranches(org, repo string) ([]github.Branch, error)
+	GetRepos(org string, user bool) ([]github.Repo, error)
+}
+
+type Protector struct {
+	client         client
+	cfg            *config.Config
+	updates        chan Requirements
```

Comment:
Why do `chan Requirements` vs `[]Requirements`? It seems to make the code harder to read without sufficient benefits (slightly faster runtime)?

Is it to work better with throttling?